tscreen_posix.go

```diff
@@ -17,150 +17,65 @@
-//     switch (cfgetospeed(tios)) {
-// #ifdef B0
-//     case B0: return (0);
-// #endif
-// #ifdef B50
-//     case B50: return (50);
-// #endif
-// #ifdef B75
-//     case B75: return (75);
-// #endif
-// #ifdef B110
-//	case B110: return (110);
-// #endif
-// #ifdef B134
-//	case B134: return (134);
-// #endif
-// #ifdef B150
-//	case B150: return (150);
-// #endif
-// #ifdef B200
-//	case B200: return (200);
-// #endif
-// #ifdef B300
-//	case B300: return (300);
-// #endif
-// #ifdef B600
-//	case B600: return (600);
-// #endif
-// #ifdef B1200
-//	case B1200: return (1200);
-// #endif
-// #ifdef B1800
-//	case B1800: return (1800);
-// #endif
-// #ifdef B2400
-//	case B2400: return (2400);
-// #endif
-// #ifdef B4800
-//	case B4800: return (4800);
-// #endif
-// #ifdef B9600
-//	case B9600: return (9600);
-// #endif
-// #ifdef B19200
-//	case B19200: return (19200);
-// #endif
-// #ifdef B38400
-//	case B38400: return (38400);
-// #endif
-// #ifdef B57600
-//	case B57600: return (57600);
-// #endif
-// #ifdef B76800
-//	case B76800: return (76800);
-// #endif
-// #ifdef B115200
-//	case B115200: return (115200);
-// #endif
-// #ifdef B153600
-//	case B153600: return (153600);
-// #endif
-// #ifdef B230400
-//	case B230400: return (230400);
-// #endif
-// #ifdef B307200
-//	case B307200: return (307200);
-// #endif
-// #ifdef B460800
-//	case B460800: return (460800);
-// #endif
-// #ifdef B921600
-//	case B921600: return (921600);
-// #endif
-//	}
-//	return (0);
-// }
-import "C"
-
-type termiosPrivate struct {
-	tios C.struct_termios
-}
+type termiosPrivate syscall.Termios
+
+var tiosp = termiosPrivate{}
 
-func (t *tScreen) termioInit() error {
-	var e error
-	var rv C.int
-	var newtios C.struct_termios
-	var fd C.int
+func (t *tScreen) termioInit() (err error) {
+	private := syscall.Termios{}
 
-	if t.in, e = os.OpenFile("/dev/tty", os.O_RDONLY, 0); e != nil {
-		goto failed
+	if t.in, err = os.OpenFile("/dev/tty", os.O_RDONLY, 0); err != nil {
+		return err
 	}
-	if t.out, e = os.OpenFile("/dev/tty", os.O_WRONLY, 0); e != nil {
-		goto failed
+	if t.out, err = os.OpenFile("/dev/tty", os.O_WRONLY, 0); err != nil {
+		return err
 	}
 
-	t.tiosp = &termiosPrivate{}
-
-	fd = C.int(t.out.Fd())
-	if rv, e = C.tcgetattr(fd, &t.tiosp.tios); rv != 0 {
-		goto failed
+	defer func() {
+		if err != nil {
+			if t.in != nil {
+				t.in.Close()
+			}
+			if t.out != nil {
+				t.out.Close()
+			}
+		}
+	}()
+
+	fd := t.out.Fd()
+
+	if err := termios.Tcgetattr(fd, &private); err != nil {
+		return fmt.Errorf("cannot get attributes: %s", err)
 	}
-	t.baud = int(C.getbaud(&t.tiosp.tios))
-	newtios = t.tiosp.tios
-	newtios.c_iflag &^= C.IGNBRK | C.BRKINT | C.PARMRK |
-		C.ISTRIP | C.INLCR | C.IGNCR |
-		C.ICRNL | C.IXON
-	newtios.c_oflag &^= C.OPOST
-	newtios.c_lflag &^= C.ECHO | C.ECHONL | C.ICANON |
-		C.ISIG | C.IEXTEN
-	newtios.c_cflag &^= C.CSIZE | C.PARENB
-	newtios.c_cflag |= C.CS8
+
+	t.baud = int(termios.Cfgetispeed(&private))
```

Comment:
Technically this should probably be Cfgetospeed, but I've yet to see a system where the serial / console input baud rate was mismatched to the output.

Reply:
Let's change it anyway. So that it is proper.